File: WdtcCore/src/main/java/org/wdt/wdtc/download/game/GameVersionList.java
package org.wdt.wdtc.download.game;


import org.apache.log4j.Logger;
import org.wdt.utils.gson.JSONArray;
import org.wdt.utils.gson.JSONObject;
import org.wdt.utils.gson.JSONUtils;
import org.wdt.wdtc.download.infterface.VersionList;
import org.wdt.wdtc.manger.FileManger;
import org.wdt.wdtc.utils.WdtcLogger;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class GameVersionList implements VersionList {
    private static final Logger logmaker = WdtcLogger.getLogger(GameVersionList.class);

    public GameVersionList() {
    }


    public List<String> getVersionList() {
        List<String> VersionList = new ArrayList<>();
        try {
            JSONArray version_list = JSONUtils.getJSONObject(FileManger.getVersionManifestFile()).getJSONArray("versions");
            for (int i = 0; i < version_list.size(); i++) {
                JSONObject VersionObject = version_list.getJSONObject(i);
                if (VersionObject.getString("type").equals("release")) {
                    VersionList.add(VersionObject.getString("id"));
                }
            }
        } catch (IOException e) {
            logmaker.error(WdtcLogger.getErrorMessage(e));
        }
        return VersionList;
    }
}

File: core/src/main/kotlin/org/wdt/wdtc/core/impl/download/game/GameVersionList.kt
package org.wdt.wdtc.core.impl.download.game

import com.google.gson.annotations.SerializedName
import org.wdt.utils.gson.getJsonArray
import org.wdt.utils.gson.readFileToJsonObject
import org.wdt.utils.io.isFileNotExists
import org.wdt.wdtc.core.openapi.download.interfaces.VersionListInterface
import org.wdt.wdtc.core.openapi.download.interfaces.VersionsJsonObjectInterface
import org.wdt.wdtc.core.openapi.download.interfaces.VersionsObjectSequence
import org.wdt.wdtc.core.openapi.download.interfaces.VersionsObjectSequence.Companion.asSequence
import org.wdt.wdtc.core.openapi.download.interfaces.VersionsObjectSequence.Companion.asVersionsObjectSequence
import org.wdt.wdtc.core.openapi.manager.versionManifestFile
import org.wdt.wdtc.core.openapi.utils.ioAsync
import java.net.URL
import java.util.*

class GameVersionList : VersionListInterface {
	
	private val list = ioAsync {
		versionManifestFile.run {
			if (isFileNotExists()) {
				DownloadVersionGameFile.startDownloadVersionManifestJsonFile()
			}
			readFileToJsonObject().getJsonArray("versions")
		}
	}
	
	override suspend fun getVersionList(): VersionsObjectSequence {
		return list.await().asSequence<GameVersionsJsonObjectImpl>().filter {
			it.gameType == "release"
		}.asVersionsObjectSequence()
	}
	
	
	class GameVersionsJsonObjectImpl(
		@field:SerializedName("id")
		override val versionNumber: String,
		
		@field:SerializedName("releaseTime")
		val releaseTime: Date,
		
		@field:SerializedName("type")
		val gameType: String,
		
		@field:SerializedName("url")
		val versionJsonURL: URL,
		
		@field:SerializedName("time")
		val time: Date
	
	) : VersionsJsonObjectInterface {
		
		
		override fun equals(other: Any?): Boolean {
			if (this === other) return true
			if (other == null || javaClass != other.javaClass) return false
			val that = other as GameVersionsJsonObjectImpl
			return versionNumber == that.versionNumber && gameType == that.gameType
		}
		
		override fun hashCode(): Int {
			return Objects.hash(versionNumber, gameType)
		}
	}
	
}